File: CPP/String/D16_Anagram.cpp
#include <iostream>
#include <string>
using namespace std;

bool areAnagrams(string &s1, string &s2)
{
      if (s1.size() != s2.size())
      {
            return false;
      }

      vector<int> freq(26, 0);

      for (size_t i = 0; i < s1.size(); i++)
      {
            freq[s1[i] - 'a']++;
            freq[s1[i] - 'a']--;
      }

      for (int count : freq)
      {
            if (count != 0)
            {
                  return false;
            }
      }

      return true;
}

int main()
{
      string s1 = "geeks", s2 = "keesg", s3 = "hello", s4 = "hell" ;
      cout << areAnagrams(s1, s2) << endl;
      cout << areAnagrams(s3, s4) << endl; 
}

File: Python/String/D16_Anagram.py
class Solution:
    def areAnagram(self, s1, s2):
        if len(s1) != len(s2):
            return False

        freq = [0] * 26

        for a, b in zip(s1, s2):
            freq[ord(a) - ord('a')] += 1
            freq[ord(b) - ord('a')] -= 1

        return all(count == 0 for count in freq)


sol = Solution()
print(sol.areAnagram("geeks", "keesg"))  
print(sol.areAnagram("hello", "hell"))  


# ord() gives you the ASCII (Unicode) integer of a character.
# zip(s1, s2)
# zip() takes two iterables and returns pairs:
# zip("abc", "xyz") → [('a','x'), ('b','y'), ('c','z')]
# You're looping over characters from both strings at once, in parallel.